minor pkg/ip fixes

bimmlerd opened the pull request:
A few small drive-by improvements to `pkg/ip`.


```
ip: remove unneeded IPTestSuite

Likely left over from a conversion from checkmate to go testing.
```
```
ip: benchmark SortAddrList

To create a baseline for the next commit.
```
```
ip: use stdlib slices package

This gives us a speedup of 30% in a microbenchmark for sorting IPs and
moves us closer to stdlib.

pkg: github.com/cilium/cilium/pkg/ip
                │     pre      │                post                 │
                │    sec/op    │   sec/op     vs base                │
SortAddrList-10   105.06µ ± 4%   72.75µ ± 1%  -30.75% (p=0.000 n=10)
```
```
slices: drop SortedUniqueFunc

The type constraints of slices.Sort and slices.Compact are different -
the former requires cmp.Ordered whereas the latter only needs
comparable. The only caller for SortedUniqueFunc was pkg/ip, for
[]netip.Addr. The usage is inefficient, though, since netip.Addr is
comparable, just not cmp.Ordered - the use of slices.CompactFunc is thus
unnecessary.

Since there's no consumers, and use can be needlessly inefficient, let's
just drop the function, even though it breaks API symmetry.
```

bimmlerd committed:
ip: remove unneeded IPTestSuite

Likely left over from a conversion from checkmate to go testing.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ip: benchmark SortAddrList

To create a baseline for the next commit.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ip: use stdlib slices package

This gives us a speedup of 30% in a microbenchmark for sorting IPs and
moves us closer to stdlib.

pkg: github.com/cilium/cilium/pkg/ip
                │     pre      │                post                 │
                │    sec/op    │   sec/op     vs base                │
SortAddrList-10   105.06µ ± 4%   72.75µ ± 1%  -30.75% (p=0.000 n=10)

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

tklauser commented on pkg/ip/ip.go:
```diff
@@ -753,15 +752,8 @@ func PartitionCIDR(targetCIDR net.IPNet, excludeCIDR net.IPNet) ([]*net.IPNet, [
 // netip.Addr.Compare (i.e. IPv4 addresses show up before IPv6).
 // The slice is manipulated in-place destructively; it does not create a new slice.
 func KeepUniqueAddrs(addrs []netip.Addr) []netip.Addr {
-	return slices.SortedUniqueFunc(
```
Looks like this is removing the only caller of `SortedUniqueFunc`. Should we remove its implementation as well given that benchmarks seem to suggest we should just sort and compact?

bimmlerd replied:
Yeah fair question - I guess for API symmetry it makes sense to also have `SortedUniqueFunc` given that we have `SortedUnique` (which is used, for better or worse). OTOH, for cases like this one, where the type is `comparable` but not `cmp.Ordered` `SortedUniqueFunc` also doesn't help because it calls `CompactFunc` instead of just `Compact`, even though netip.Addrs are `comparable`.

Given the two different type constraints I'll drop it I think.

(I didn't benchmark `KeepUniqueAddrs` only the sorting funcs.)

sayboras commented on pkg/ip/ip_test.go:
```diff
@@ -13,12 +13,6 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-type IPTestSuite struct{}
```
thanks 👍

bimmlerd committed:
slices: drop SortedUniqueFunc

The type constraints of slices.Sort and slices.Compact are different -
the former requires cmp.Ordered whereas the latter only needs
comparable. The only caller for SortedUniqueFunc was pkg/ip, for
[]netip.Addr. The usage is inefficient, though, since netip.Addr is
comparable, just not cmp.Ordered - the use of slices.CompactFunc is thus
unnecessary.

Since there's no consumers, and use can be needlessly inefficient, let's
just drop the function, even though it breaks API symmetry.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

ovidiutirla reviewed (approved):
Looking at the performance benchmark and the number of usages, agree to remove the `SortedUniqueFunc`.

sayboras reviewed (approved):
Thanks
